### PR TITLE
Add Terraform MajorCloud Providers

### PR DESCRIPTION
**Pull request by sodre**
For versions >=0.10 terraform moved to a plugin-based system for its resource providers. 
This PR adds the MajorCloud plugins as separate conda-packages. Additional PRs will be started for the additional plugins. There are quite a few number of plugins so it is easier for maintenance to simply parse the SHA256SUM file from the upstream provider. 

Windows Builds:
 - Until hashicorp/terraform#16917 and golang/go#18468 are closed we have to take the upstream precompiled binaries or not distribute the Windows binaries.

Lint Issue:
 - Parsing the SHA256SUM is done using the load_file_regex command from conda-build, which is not understood by the linter yet. See conda-forge/conda-smithy#707 and conda-forge/conda-smithy#708 for the linter PR.

### CODE REVIEW

**Comment by sodre** [comment]
@isuruf 
When you get a chance can you review the terraform-provider series of recipes? They are ready to go from my side with the exception of the linter issue we took care of in conda-smithy 3.x series.

**Comment by sodre** [comment]
@jakirkham, would you mind taking a look at the Terraform-Provider recipes and give me your input. There are quite a few of them, but they are all based on the same recipe. If you're okay with it It makes it easier to merge when the other PR's checks pass again. (I had to rebase because of the AppVeyor issue).

**Comment by sodre** [comment]
@ocefpaf, could you take a quick look at the Terraform-platform PRs? I rewrote the recipe so it does not affect the regro-tick-bot.

File: recipes/terraform-provider-alicloud/meta.yaml [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+  url: https://releases.hashicorp.com/{{ name }}/{{ version }}/{{ name }}_{{ version }}_windows_386.zip  # [win32]
+  url: https://releases.hashicorp.com/{{ name }}/{{ version }}/{{ name }}_{{ version }}_windows_amd64.zip  # [win64]
+  fn: {{ name }}-{{ version }}.zip  # [win]
+  sha256: {{ sha256_win32 }}  # [win32]
+  sha256: {{ sha256_win64 }}  # [win64]
+
+build:
+  number: 0
+  script:
+    - XCOPY {{ x4_plugin }}.exe %LIBRARY_BIN%  # [win]
```
Comment (isuruf) [inline recipes/terraform-provider-alicloud/meta.yaml]: Should the exe be versioned?

File: recipes/terraform-provider-aws/build.sh [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+# Link code to GOPATH directory
+ln -s $(pwd) $GOPATH/src/github.com/terraform-providers/$PKG_NAME
+
+# Build
+cd $GOPATH/src/github.com/terraform-providers/$PKG_NAME
+make build
+make test
+
+# Install Binary into PREFIX/bin
+mv $GOPATH/bin/$PKG_NAME $PREFIX/bin/${PKG_NAME}_v${PKG_VERSION}_x4
```
Comment (isuruf) [inline recipes/terraform-provider-aws/build.sh]: Same here

File: recipes/terraform-provider-alicloud/meta.yaml (hunk 1 above)
Comment (sodre) [reply recipes/terraform-provider-alicloud/meta.yaml]: the short answer is yes. I will follow up with the link to the upstream documentation in the next post.

**Comment by sodre** [comment]
Excerpt from https://www.hashicorp.com/blog/writing-custom-terraform-providers

```
Next, build the plugin using the Go toolchain:

$ go build -o terraform-provider-example
The output name (-o) is very important. Terraform searches for plugins in the format of:

terraform-<TYPE>-<NAME>
In the case above, the plugin is of type "provider" and of name "example".
```

The x4 is needed because these are not "third-party" plugins, i.e. they are getting installed side-by-side to the main terraform app.

**Comment by sodre** [comment]
@isuruf,

Here is the language that I was looking for from [Terraform 0.11 Docs](https://www.terraform.io/docs/configuration/providers.html):

> The naming scheme for provider plugins is terraform-provider-NAME_vX.Y.Z, and Terraform uses the name to understand the name and version of a particular provider binary. Third-party plugins will often be distributed with an appropriate filename already set in the distribution archive so that it can be extracted directly into the plugin directory described above.

There is also an entry on StackOverflow, https://stackoverflow.com/questions/47603344/terraform-plugin-development-where-to-specify-version-best-practices-for-go-ap

Finally, the 'test" section actually checks that the naming convention works. 

If the explanation is sufficient, can you merge the whole set of PRs?

**Comment by sodre** [comment]
Thanks @isuruf, can you take care of the other ones as well ? #5329, #5330, #5331 #5332, #5333, #5334, #5335 ?

**Pull request merged**